velox/functions/sparksql/RegexFunctions.cpp

```diff
@@ -57,6 +81,119 @@ void ensureRegexIsCompatible(
+/// regex_replace(string, pattern, overwrite, position) → string
+///
+/// If a string has a substring that matches the given pattern, replace
+/// the match in the string wither overwrite and return the string. If
+/// optional paramter position is provided, only make replacements
+/// after that positon in the string (1 indexed).
+///
+/// If position <= 0, throw error.
+/// If position > length string, return string.
+template <typename T>
+struct RegexReplaceFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace) {
+    checkForCompatiblePattern(pattern);
+
+    re2::RE2* patternRegex = getCachedRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+
+    std::string string(stringInput.data(), stringInput.size());
+    RE2::GlobalReplace(&string, *patternRegex, replaceStringPiece);
+
+    // Prepare the final result string without unnecessary copying
+    if (string.size()) {
+      result.resize(string.size());
+      std::memcpy(result.data(), string.data(), string.size());
+    }
+  }
+
+  void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& stringInput,
+      const arg_type<Varchar>& pattern,
+      const arg_type<Varchar>& replace,
+      const arg_type<int64_t>& position) {
+    VELOX_CHECK_LT(-1, position - 1);
+    checkForCompatiblePattern(pattern);
+
+    re2::RE2* patternRegex = getCachedRegex(pattern.str());
+    re2::StringPiece replaceStringPiece = toStringPiece(replace);
+    re2::StringPiece inputStringPiece(stringInput.data(), stringInput.size());
+
+    const char* prefixEnd = inputStringPiece.data() +
+        std::min(stringInput.size(), static_cast<size_t>(position - 1));
+    re2::StringPiece prefix(
+        inputStringPiece.data(), prefixEnd - inputStringPiece.data());
+
+    re2::StringPiece targetStringPiece(
+        prefixEnd, inputStringPiece.end() - prefixEnd);
+
+    std::string targetString(
+        targetStringPiece.data(), targetStringPiece.size());
+    RE2::GlobalReplace(&targetString, *patternRegex, replaceStringPiece);
+
+    // Prepare the final result string without unnecessary copying
+    if (targetString.size() || prefix.size()) {
+      result.resize(prefix.size() + targetString.size());
+      std::memcpy(result.data(), prefix.data(), prefix.size());
+      std::memcpy(
+          result.data() + prefix.size(),
+          targetString.data(),
+          targetString.size());
+    }
+  }
+
+ private:
+  re2::RE2* getCachedRegex(const std::string& pattern) const {
+    auto it = patternCache_.find(pattern);
+    if (it != patternCache_.end()) {
+      return it->second.get();
+    }
+    VELOX_USER_CHECK_LT(
+        patternCache_.size(),
+        kMaxCompiledRegexes,
+        "regex_replace hit the maximum number of unique regexes: kMaxCompiledRegexes");
```

Comment:
The user would not know what kMaxCompiledRegexes is. Please, use the actual value.